### ISSUE TITLE
Assertion failed: S_finalize_op (op.c:2562)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14927]
Migrated from [rt.perl.org#126170](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126170) (status was 'open')

Searchable as RT126170$

**Comment by p5pRT** [comment]
### From @geeknik

Fuzzing perl v5\.23\.4 \(v5\.23\.3\-7\-ge120c24\) with AFL found the following assertion failure&#8203;:

~/perl/perl \-e 'BEGIN\(\)&#8203;:o'

Prototype mismatch&#8203;: sub main&#8203;::BEGIN \(\) vs none at \-e line 1\.
perl&#8203;: op\.c&#8203;:2562&#8203;: S\_finalize\_op&#8203;: Assertion \`has\_last || family == \(1 \<\< 8\) || family == \(15 \<\< 8\) || family == \(3 \<\< 8\) || family == \(11 \<\< 8\) || family == \(12 \<\< 8\) || family == \(13 \<\< 8\) || family == \(14 \<\< 8\) || type == OP\_SASSIGN || type == OP\_CUSTOM || type == OP\_NULL' failed\.
Aborted

gdb output&#8203;:
Prototype mismatch&#8203;: sub main&#8203;::BEGIN \(\) vs none at test00 line 1\.
perl&#8203;: op\.c&#8203;:2562&#8203;: S\_finalize\_op&#8203;: Assertion \`has\_last || family == \(1 \<\< 8\) || family == \(15 \<\< 8\) || family == \(3 \<\< 8\) || family == \(11 \<\< 8\) || family == \(12 \<\< 8\) || family == \(13 \<\< 8\) || family == \(14 \<\< 8\) || type == OP\_SASSIGN || type == OP\_CUSTOM || type == OP\_NULL' failed\.

Program received signal SIGABRT\, Aborted\.
0x00007ffff6d90165 in \*\_\_GI\_raise \(sig=\<optimized out>\)
&nbsp;    at \.\./nptl/sysdeps/unix/sysv/linux/raise\.c&#8203;:64
64    \.\./nptl/sysdeps/unix/sysv/linux/raise\.c&#8203;: No such file or directory\.
\(gdb\) bt
\#0  0x00007ffff6d90165 in \*\_\_GI\_raise \(sig=\<optimized out>\)
&nbsp;    at \.\./nptl/sysdeps/unix/sysv/linux/raise\.c&#8203;:64
\#1  0x00007ffff6d933e0 in \*\_\_GI\_abort \(\) at abort\.c&#8203;:92
\#2  0x00007ffff6d89311 in \*\_\_GI\_\_\_assert\_fail \(
&nbsp;    assertion=0xef41c8 "has\_last || family == \(1 \<\< 8\) || family == \(15 \<\< 8\) || family == \(3 \<\< 8\) || family == \(11 \<\< 8\) || family == \(12 \<\< 8\) || family == \(13 \<\< 8\) || family == \(14 \<\< 8\) || type == OP\_SASSIGN || type =="\.\.\.\, file=\<optimized out>\, line=2562\, 
&nbsp;    function=0xf08387 "S\_finalize\_op"\) at assert\.c&#8203;:81
\#3  0x0000000000443cd1 in S\_finalize\_op \(\) at op\.c&#8203;:2549
\#4  0x0000000000443f90 in S\_finalize\_op \(\) at op\.c&#8203;:2579
\#5  0x0000000000443f90 in S\_finalize\_op \(\) at op\.c&#8203;:2579
\#6  0x0000000000443f90 in S\_finalize\_op \(\) at op\.c&#8203;:2579
\#7  0x00000000004e0432 in Perl\_newATTRSUB\_x \(\) at op\.c&#8203;:2381
\#8  0x00000000004f5459 in Perl\_utilize \(\) at op\.c&#8203;:6041
\#9  0x00000000004f7a14 in Perl\_load\_module\.constprop\.43 \(\) at op\.c&#8203;:6188
\#10 0x00000000004db764 in Perl\_newATTRSUB\_x \(\) at op\.c&#8203;:8631
\#11 0x000000000067332c in Perl\_yyparse \(\)
\#12 0x000000000053c005 in S\_parse\_body \(\) at perl\.c&#8203;:2304
\#13 0x0000000000543d83 in perl\_parse \(\) at perl\.c&#8203;:1634
\#14 0x000000000042c6f8 in main \(\) at perlmain\.c&#8203;:114

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Sep 24\, 2015 at 07&#8203;:41&#8203;:54PM \-0700\, Brian Carpenter wrote&#8203;:
> Fuzzing perl v5\.23\.4 \(v5\.23\.3\-7\-ge120c24\) with AFL found the following assertion failure&#8203;:
> 
> ~/perl/perl \-e 'BEGIN\(\)&#8203;:o'
> 
> Prototype mismatch&#8203;: sub main&#8203;::BEGIN \(\) vs none at \-e line 1\.
> perl&#8203;: op\.c&#8203;:2562&#8203;: S\_finalize\_op&#8203;: Assertion \`has\_last || family == \(1 \<\< 8\) || family == \(15 \<\< 8\) || family == \(3 \<\< 8\) || family == \(11 \<\< 8\) || family == \(12 \<\< 8\) || family == \(13 \<\< 8\) || family == \(14 \<\< 8\) || type == OP\_SASSIGN || type == OP\_CUSTOM || type == OP\_NULL' failed\.

This is occurring basically because toward the end of compiling BEGIN\,
its tries to apply the attribute "o"\, which triggers an attempt to
"use attributes"\, which results in faking up a BEGIN block\, which when its
finished being compiled\, causes the 'old' BEGIN to be freed\, which
triggers freeing any unattached ops associated with the old BEGIN's
compilation\, which prematurely frees the ops used to construct the
import\(\) call\, which eventually leads to the assert failure\.

This only happens because the attributue application happens at a specific
point during the BEGIN's compilation\, where is hasn't yet had the root of
its optree attached to CvROOT\. When it's subsequently freed\, it looks to
cv\_undef\_flags\(\) like a sub compilation that's errored out\, and thus whose
op slab needs freeing\, freeing up any orphaned ops\. This is why

&nbsp;    BEGIN \{ BEGIN \{\} \}
and
&nbsp;    BEGIN \{\} BEGIN \{\}

don't trigger the same issue\.

The whole mess in newATTRSUB\_x\(\) is beyond my understanding\. I think it
might just be easier to croak if someone tries to apply attributes to a
BEGIN block \(or any other special BLOCK?\)

Can anyone think or a reason why special blocks should be allowed
attributes\, or whether such usage exists in the wild?


\-\- 
This email is confidential\, and now that you have read it you are legally
obliged to shoot yourself\. Or shoot a lawyer\, if you prefer\. If you have
received this email in error\, place it in its original wrapping and return
for a full refund\. By opening this email\, you accept that Elvis lives\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ap

\* Dave Mitchell \<davem@&#8203;iabyn\.com> \[2015\-10\-08 16&#8203;:55\]&#8203;:
> or whether such usage exists in the wild?

Fiddling around with grep\.cpan\.me I found no use on CPAN whatsoever\.

There is some very minor use of \`sub BEGIN\` and some even more minor use
of the empty prototype on BEGIN blocks\, but I did not find even so much
as an empty attribute list on any of them\, let alone actual attributes\.

> Can anyone think or a reason why special blocks should be allowed
> attributes

I think it should be an error\. But that’s in a vacuum\. It has been
possible to write them\, for a long time\, so the question is whether we
gain anything but outlawing them now\. The obvious answer is we fix the
assert crash\. But wait\, there’s an assert crash\, so \*can\* this syntax
even work at all? If not\, then we can trivially outlaw it because it
doesn’t work and can’t be used anyway\. If yes\, the question is again\,
given the circumstances in which it works\, their likelihood\, etc\, what
do we gain from outlawing it?

My expectation is that “just kill it” is not particularly questionable\,
but I don’t feel I am firm enough on the answers to actually take that
stance\.

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
### From @arc

Aristotle Pagaltzis \<pagaltzis@&#8203;gmx\.de> wrote&#8203;:
> \* Dave Mitchell \<davem@&#8203;iabyn\.com> \[2015\-10\-08 16&#8203;:55\]&#8203;:
>> Can anyone think or a reason why special blocks should be allowed
>> attributes
>
> I think it should be an error\. But that’s in a vacuum\. It has been
> possible to write them\, for a long time\, so the question is whether we
> gain anything but outlawing them now\. The obvious answer is we fix the
> assert crash\. But wait\, there’s an assert crash\, so \*can\* this syntax
> even work at all?

This attempt&#8203;:

sub MODIFY\_CODE\_ATTRIBUTES \{ \(\) \} sub BEGIN :Foo \{\}

does nothing \(while emitting no warnings\) under 5\.6 through 5\.14\. As
of 5\.16\, it says&#8203;:

Subroutine BEGIN redefined at \-e line 1\.
Attempt to free unreferenced scalar&#8203;: SV 0xdeadbeef at \-e line 1\.

but the code still executes\. \(I haven't tried to work out how much of
a problem the "unreferenced scalar" error is\.\)

Removing the "sub" keyword from the attribute\-attached BEGIN&#8203;:

sub MODIFY\_CODE\_ATTRIBUTES \{ \(\) \} BEGIN :Foo \{\}

produces this error \(or a reworded equivalent as of 5\.18\) on all Perls
from 5\.6 onwards&#8203;:

Can't call method "Foo" on an undefined value at \-e line 1\.

because it's actually being parsed as the rough equivalent of

BEGIN&#8203;: \(do \{\}\)\->Foo

— that is\, a statement with the label "BEGIN"\, in which the "Foo"
method is invoked \(using indirect\-object notation\) on the result of an
empty block\.

To summarise&#8203;: in the past we've had the ability to attach attributes
to "sub BEGIN"\, but nobody noticed when we apparently\-accidentally
broke some aspects of it in 5\.16; and we've never had the ability to
attach attributes to "BEGIN" without "sub"\.

> If not\, then we can trivially outlaw it because it
> doesn’t work and can’t be used anyway\.

It seems to me that outlawing it would indeed be safe\. But we
presumably need tests that "BEGIN : stuff" treats the stuff as a
labelled statement\.

\-\- 
Aaron Crane \*\* http&#8203;://aaroncrane\.co\.uk/
